### PR TITLE
Create test runner

### PR DESCRIPTION
**Pull request by lpil**
Hello!

Thank you for making such a wonderful template repository. It was very easy to create a Gleam test runner using it as a base.

Closes https://github.com/exercism/gleam/issues/28

### CODE REVIEW

File: bin/run.sh [hunk 1]
```diff
@@ -17,44 +17,66 @@
...
+echo "Copying dependencies..."
+cd packages
+mkdir -p "$solution_dir"/build
+cp -r "$root_dir"/packages/build/packages "$solution_dir"/build/packages
+
+sanitise_gleam_output() {
+  grep -vE \
+    -e "^Downloading packages" \
+    -e "^ Downloaded [0-9]+ packages in [0-9]\.[0-9]+s" \
+    -e "^ Downloaded [0-9]+ packages in [0-9]\.[0-9]+s" \
```
Comment (michallepicki) [inline bin/run.sh:41]: Will these lines be printed anyway even if dependencies were copied from the packages project?

Comment (lpil) [reply bin/run.sh:41]: I don't think so, no. This is likely reundant now, I'll remove it

Comment (lpil) [reply bin/run.sh:41]: I was wrong, it may still get printed

**Review by ErikSchierboom — APPROVED** [review]
LGTM! One thing I did find is that running the tests seemed to be relatively slow, and I have a beefy machine. We'll need to see if there are timeouts once we have this deployed.

> Thank you for making such a wonderful template repository. It was very easy to create a Gleam test runner using it as a base.

Thanks! This is lovely to hear.

Note that for the test runner to be deployed, @iHiD needs to merge https://github.com/exercism/terraform/pull/82 and then run terraform (as well as giving proper access to the right secrets).

File: README.md [hunk 2]
```diff
@@ -44,7 +27,8 @@ Once the test runner has finished, its results will be written to `<output-dir>/
 To run the tests to verify the behavior of the test runner, do the following:
 
 1. Open a terminal in the project's root
-2. Run `./bin/run-tests.sh`
+2. Run `./bin/setup-locally.sh`
+3. Run `./bin/run-tests.sh`
```
Comment (chouzar) [inline README.md:31]: Running step 2 I get: 

```
➜  gleam-test-runner git:(create) ./bin/setup-locally.sh
Downloading packages
 Downloaded 2 packages in 0.84s
```

But when trying step 3 the `jq` dependency is needed:

```
➜  gleam-test-runner git:(create) ./bin/run-tests.sh
all_fail: testing...
bin/run.sh: line 78: jq: command not found
```

Just wondering if this is the expected behaviour.

Comment (chouzar) [reply README.md:31]: When running tests using Docker I get this output: 

```
➜  gleam-test-runner git:(create) ./bin/run-tests-in-docker.sh
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
all_fail: testing...
--- /opt/test-runner/tests/all_fail/results.json
+++ /opt/test-runner/tests/all_fail/expected_results.json
@@ -1,5 +1,5 @@
 {
   "version": 1,
-  "status": "error",
-  "message": "error: HTTP error\n\nA HTTP request failed.\nThe error from the HTTP client was:\n\n    error sending request for url (https://repo.hex.pm/tarballs/gleam_stdlib-0.24.0.tar): error trying to connect: dns error: failed to lookup address information: Try again"
+  "status": "fail",
+  "message": "\u001b[0;31mF\u001b[0m\u001b[0;31mF\u001b[0m\nFailures:\n\n  1) all_fail_test:one_test/0\n\u001b[0;31m     Failure: ?assertEqual(<<111,111,112,115>>, Actual)\n       expected: <<\"oops\">>\n            got: <<\"one\">>\u001b[0m\n\u001b[0;36m     %% eunit_proc.erl:346:in `eunit_proc:with_timeout/3`\u001b[0m\n\u001b[0;36m     Output: \u001b[0m\n\u001b[0;36m     Output: \u001b[0m\n\n  2) all_fail_test:two_test/0\n\u001b[0;31m     Failure: ?assertEqual(<<111,111,112,115>>, Actual)\n       expected: <<\"oops\">>\n            got: <<\"two\">>\u001b[0m\n\u001b[0;36m     %% eunit_proc.erl:346:in `eunit_proc:with_timeout/3`\u001b[0m\n\u001b[0;36m     Output: \u001b[0m\n\u001b[0;36m     Output: \u001b[0m\n\n\u001b[0;31m2 tests, 2 failures\n\u001b[0m"
 }

all_fail: fail.
all_fail: /opt/test-runner/tests/all_fail/results.json does not match /opt/test-runner/tests/all_fail/expected_results.json

empty_file: testing...
--- /opt/test-runner/tests/empty_file/results.json
+++ /opt/test-runner/tests/empty_file/expected_results.json
@@ -1,5 +1,5 @@
 {
   "version": 1,
   "status": "error",
-  "message": "error: HTTP error\n\nA HTTP request failed.\nThe error from the HTTP client was:\n\n    error sending request for url (https://repo.hex.pm/tarballs/gleam_stdlib-0.24.0.tar): error trying to connect: dns error: failed to lookup address information: Try again"
+  "message": "error: Unknown module field\n   ┌─ ./test/empty_file_test.gleam:10:13\n   │\n10 │   empty_file.hello_world()\n   │             ^^^^^^^^^^^^\n\nThe module `empty_file` does not have a `hello_world` field."
 }

empty_file: fail.
empty_file: /opt/test-runner/tests/empty_file/results.json does not match /opt/test-runner/tests/empty_file/expected_results.json

partial_fail: testing...
--- /opt/test-runner/tests/partial_fail/results.json
+++ /opt/test-runner/tests/partial_fail/expected_results.json
@@ -1,5 +1,5 @@
 {
   "version": 1,
-  "status": "error",
-  "message": "error: HTTP error\n\nA HTTP request failed.\nThe error from the HTTP client was:\n\n    error sending request for url (https://repo.hex.pm/tarballs/gleam_stdlib-0.24.0.tar): error trying to connect: dns error: failed to lookup address information: Try again"
+  "status": "fail",
+  "message": "\u001b[0;32m.\u001b[0m\u001b[0;32m.\u001b[0m\u001b[0;31mF\u001b[0m\nFailures:\n\n  1) partial_fail_test:three_test/0\n\u001b[0;31m     Failure: ?assertEqual(<<111,111,112,115>>, Actual)\n       expected: <<\"oops\">>\n            got: <<\"three\">>\u001b[0m\n\u001b[0;36m     %% eunit_proc.erl:346:in `eunit_proc:with_timeout/3`\u001b[0m\n\u001b[0;36m     Output: \u001b[0m\n\u001b[0;36m     Output: \u001b[0m\n\n\u001b[0;31m3 tests, 1 failures\n\u001b[0m"
 }

partial_fail: fail.
partial_fail: /opt/test-runner/tests/partial_fail/results.json does not match /opt/test-runner/tests/partial_fail/expected_results.json

success: testing...
--- /opt/test-runner/tests/success/results.json
+++ /opt/test-runner/tests/success/expected_results.json
@@ -1,5 +1,4 @@
 {
   "version": 1,
-  "status": "error",
-  "message": "error: HTTP error\n\nA HTTP request failed.\nThe error from the HTTP client was:\n\n    error sending request for url (https://repo.hex.pm/tarballs/gleam_stdlib-0.24.0.tar): error trying to connect: dns error: failed to lookup address information: Try again"
+  "status": "pass"
 }

success: fail.
success: /opt/test-runner/tests/success/results.json does not match /opt/test-runner/tests/success/expected_results.json

syntax_error: testing...
--- /opt/test-runner/tests/syntax_error/results.json
+++ /opt/test-runner/tests/syntax_error/expected_results.json
@@ -1,5 +1,5 @@
 {
   "version": 1,
   "status": "error",
-  "message": "error: HTTP error\n\nA HTTP request failed.\nThe error from the HTTP client was:\n\n    error sending request for url (https://repo.hex.pm/tarballs/gleam_stdlib-0.24.0.tar): error trying to connect: dns error: failed to lookup address information: Try again"
+  "message": "error: Syntax error\n  ┌─ ./src/syntax_error.gleam:5:1\n  │\n5 │ }\n  │ ^ I was not expecting this.\n\nExpected one of: \"]\""
 }

syntax_error: fail.
syntax_error: /opt/test-runner/tests/syntax_error/results.json does not match /opt/test-runner/tests/syntax_error/expected_results.json

Some tests failed!
```

A bit puzzled by the last message `Some tests failed!`.

**Review by chouzar — COMMENTED** [review]
Tried to do some tests but I'm likely missing something.

File: README.md (hunk 2 above)
Comment (ErikSchierboom) [reply README.md:31]: > The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested

This is due to you running an arm machine (maybe a MacBook) and the image being compiled for amd64

**Comment by lpil** [comment]
> LGTM! One thing I did find is that running the tests seemed to be relatively slow, and I have a beefy machine. We'll need to see if there are timeouts once we have this deployed.

@ErikSchierboom That's interesting- how long did they take to run for you? It takes 2 seconds to run an exercise for me.

File: README.md (hunk 2 above)
Comment (lpil) [reply README.md:31]: I've added notes on what needs to be installed, thanks!

Comment (lpil) [reply README.md:33]: I've added notes on what needs to be installed, thanks!

**Comment by lpil** [comment]
I've attempted to improve performance by precompiling the deps. I could possibly optimise it further by using the JavaScript backend, but seeing as there's lots of tracks that compile a lot slower than Erlang (and a literal Erlang track!) that shouldn't be required. Perhaps there is some other problem?

File: README.md [hunk 3]
```diff
@@ -1,21 +1,4 @@
...
-   - Update the files to match your track's needs. At the very least, you'll need to update `bin/run.sh`, `Dockerfile` and the test solutions in the `tests` directory
-   - Tip: look for `TODO:` comments to point you towards code that need updating
-   - Tip: look for `OPTIONAL:` comments to point you towards code that _could_ be useful
-
-Once you're happy with your test runner, [open an issue on the exercism/exercism](https://github.com/exercism/exercism/issues/new?assignees=&labels=&template=new-test-runner.md&title=%5BNew+Test+Runner%5D+) to request an official test runner repository for your track.
-
-# Exercism TRACK_NAME_HERE Test Runner
+# Exercism Gleam Test Runner
 
 The Docker image to automatically run tests on TRACK_NAME_HERE solutions submitted to [Exercism].
```
[the remark this answers is not in the document]
Comment (lpil) [reply README.md]: Thanks

**Comment by lpil** [comment]
OK, thank you!

**Pull request merged**